src/machine/shared_library.rs

```diff
@@ -0,0 +1,269 @@
+        /// not be deallocated until the program terminates.
+        ///
+        #[export_name = "scryer_machine_free"]
+        pub unsafe extern "C" fn machine_free(ptr: *mut Machine) {
+            unsafe {
+                drop(Box::from_raw(ptr));
+            }
+        }
+
+        /// Returns a new query generator for the given virtual machine.
+        ///
+        /// # Safety
+        ///
+        /// Caller must satisfy the following preconditions for this function:
+        ///
+        /// * Valid [`Machine`] pointer created with [`machine_new`] that has not yet been freed.
+        /// * the input fulfill the safety requirements of [`CStr::From_ptr`].
+        /// * There must be no other [`QueryState`] for this [`Machine`] started by
+        /// [`run_query_iter`] that has not yet been freed with
+        /// [`query_state_free`] or the [`Machine`] state will enter an undefined
+        /// configuration with unpredictable results.
+        ///
+        /// Other concerns:
+        /// * after invoking this function, calling any other function besides [`run_query_next`]
+        /// before invoking [`query_state_free`] on the [`Machine`] pointer and [`QueryState`] pointer
+        /// will leave the [`Machine`] in an undefined state
+        ///
+        #[export_name = "scryer_run_query_iter"]
+        pub unsafe extern "C" fn run_query_iter(
+            machine: &mut Machine,
+            input: *const c_char,
+        ) -> *mut QueryState {
+            match unsafe { CStr::from_ptr(input) }.to_str() {
+                Ok(input) => {
+                    let string = input.to_string();
+                    let query_state = machine.run_query_iter(string);
+                    Box::into_raw(Box::new(query_state))
+                }
+                Err(err) => {
+                    eprintln!("Query input wasn't valid UTF-8: {err:?}");
+                    std::ptr::null_mut()
+                }
+            }
+        }
+
+        /// Cleans up the [`QueryState`] in the associated Scryer [`Machine`].
+        ///
+        /// # Safety
+        ///
+        /// * `query_state` must be a valid mutable pointer to a [`QueryState`] created by [`run_query_iter`]
+        /// that has not yet been freed by [`query_state_free`].
+        /// * There can be only one [`QueryState`] per [`Machine`] started by
+        /// [`run_query_iter`] that has not yet been freed with
+        /// [`query_state_free`] or the [`Machine`] state will enter an undefined
+        /// configuration with unpredictable results.
+        #[export_name = "scryer_query_state_free"]
+        pub unsafe extern "C" fn query_state_free(query_state: *mut QueryState) {
+            unsafe {
+                drop(Box::from_raw(query_state));
+            }
+        }
+
+        /// Returns a NULL POINTER if no addition iterations, else returns a
+        /// JSON encoded UTF-8 string with one iteration of results from a Scryer Prolog query.
```

Comment:
"UTF-8 encoded JSON string" is better probably. The text encoding is UTF-8, the serialization format is JSON.